neps/nep-0424.md

```diff
@@ -0,0 +1,314 @@
+4 * read_cached_trie_node
+```
+
+Thanks to Bob's request earlier, the first four trie nodes Boormika touches are
+in the cache. Reading them is charged at a discount. (`read_cached_trie_node` is
+currently about 8 times cheaper than `touching_trie_node`.)
+
+Boormika is happy now but she will be unpleasantly surprised when she pays more
+next time because Bob was not sending his transaction.
+
+
+## Reference Implementation
+
+No code changes are necessary. Setting these two parameters to zero already
+satisfies the main premise of this NEP.
+
+```
+wasm_touching_trie_node:       16_101_955_926 ->    0
+wasm_read_cached_trie_node:     2_280_000_000 ->    0
+```
+
+However, some existing parameters need to be increased in order to keep gas
+costs balanced with the time a client needs to fulfill storage operations.
+
+The reference implementation uses parameter values that are in principle
+feasible today, both from a dApp developer and a nearcore developer perspective.
+However, the exact numbers are an implementation detail that will likely change
+over time, as hardware and client implementation become faster.
+
+```
+wasm_storage_read_base:        56_356_845_750 -> 200_000_000_000
+wasm_storage_has_key_base:     54_039_896_625 -> 200_000_000_000
+wasm_storage_write_base:       64_196_736_000 -> 220_000_000_000
+wasm_storage_remove_base:      53_473_030_500 -> 220_000_000_000
+wasm_storage_read_key_byte:    30_952_533     ->     100_000_000
+wasm_storage_has_key_byte:     30_790_845     ->     100_000_000
+wasm_storage_write_key_byte:   70_482_867     ->   1_000_000_000
+wasm_storage_remove_key_byte:  38_220_384     ->   1_000_000_000
+```
+
+The next section summarizes the rational behind these changes.
+### Increasing Existing Parameter Values
+
+The main factors to consider are:
+1. How will a changed gas cost affect current dApps on NEAR?
+2. What gas costs are required to guarantee system stability.
+
+A series of experiments have been conducted to explore the parameter space and
+evaluate its impact on existing dApps. The results for how users are affected is
+summaries in
+[Eliminate TTN gas cost](https://docs.google.com/document/d/1QoHX0xnPg0zlBJAMyEBdryx_JMuS2RzygBJ-qgRJA0Y/).
+The nearcore performance evaluation still needs to finish. (TODO: link it here)
+
+The shortest possible summary is this.
+- Base costs go to 200 Ggas to cover two disk accesses. Today's usage patterns
+  are largely unaffected by this.
+- Read costs per key byte can stay low enough that nobody should notice it.
+- Write per key byte costs need to be increased to at least 1 Ggas. This is
+  noticeable to users but should be acceptable.
+
+TODO: Include examples for how users are affected by the change.
+
+## Deploying the Feature
+
+Unfortunately, nearcore is not ready yet to support such low gas costs for
+writes. For now, it still has to do an expensive trie traversal for every write
+before it continues smart contract execution. This will be optimized in the
```

Comment:
I'm a little bit surprised by this statement - as in theory, this traversal is as expensive as the one during read, right?

Reply:
Yes, traversal costs the same for read and write. This is based on the assumption that flat storage for read is complete and for write is is not.

We have some details left to do for reads as well but it will be ready for the same release that this proposal could possibly be included anyway. For writes, it will be significantly more work, so we want to keep doing the expensive traversal for writes.